### PR TITLE
Data handling

### PR DESCRIPTION
**Pull request by toml01**

### CODE REVIEW

**Review by liorbond — APPROVED** [review]

File: development/secret-contracts/best-practices/data-handling.md [hunk 1]
```diff
@@ -2,14 +2,82 @@
...
+More specifically, since `Uint128` is a string encoded number the storage space it'll consume will depend on the number of digits of the number you are storing. `u128` on the other hand will always take a constant amount of storage space. That's why `Uint128` will be more efficient for very small numbers (and then, why use 128-bit integer to begin with?), while `u128` will be more efficient for most use cases.
+
+Example:
+
+```rust
+let n1: Uint128 = Uint128::new(10); // 2 bytes
+let n2: u128 = 10;                  // 4 bytes
+
+let n3: Uint128 = Uint128::new(12345678); // 8 bytes
+let n4: u128 = 10;                        // 4 bytes
```
Comment (eshelB) [inline development/secret-contracts/best-practices/data-handling.md:48]: maybe compare the same number, 12345678?

File: development/secret-contracts/best-practices/data-handling.md [hunk 2]
```diff
@@ -2,14 +2,82 @@
...
+
+Sometimes you can absorb some lack of precision, and you can use integer division. For example, if you want to divide 1 million tokens between three addresses:
+
+```rust
+let to_divide: u128 = 1_000_000;
+
+let addr_a: u128 = to_divide / 3; // 333,333
+let addr_b: u128 = to_divide / 3; // 333,333
+let addr_c: u128 = to_divide / 3 + 1; // 333,334
+```
```
Comment (eshelB) [inline development/secret-contracts/best-practices/data-handling.md:67]: maybe another method to workaround floats: SCALE_FACTOR? as we do with the staking/tax contracts?

File: development/secret-contracts/best-practices/data-handling.md [hunk 3]
```diff
@@ -2,14 +2,82 @@
...
-### How Uint128s Can Become A Security Risk&#x20;
+Simliarly, `cosmwasm-std` also has `Uint64` and `Uint256` and all of the following applies there as well.
+
+### When To Use Uint128 Instead Of u128
+
+`Uint128` is a thin wrapper around `u128` that is using strings for JSON encoding/decoding, such that the full `u128` range can be used for clients that convert JSON numbers to floats, like JavaScript and jq ([source](https://docs.cosmwasm.com/docs/1.0/smart-contracts/math/#uint128)).
+
+### Entrypoint Messages
+
+If you are familiar with Messages, you already know that most of the time we will use [serde](https://serde.rs/) to deserialize them from JSON (if not, you should read on [contract entrypoints](https://docs.cosmwasm.com/docs/1.0/actor-model/actor-in-blokchain/#entry-points) and the concept of [Messages](https://docs.cosmwasm.com/docs/1.0/smart-contracts/message/message)). Output will often be serizlized in the same way.
```
Comment (eshelB) [inline development/secret-contracts/best-practices/data-handling.md:17]: typo: serizlized

File: development/secret-contracts/best-practices/data-handling.md [hunk 4]
```diff
@@ -2,14 +2,82 @@
...
 
-### Data Hygiene&#x20;
+If you are familiar with Rust, you might know that it has it's own native primitive - `u128`. `Uint128` differs from `u128` in that it is a string encoded integer, rather than the traditional little/big-endian.
 
-### How Uint128s Can Become A Security Risk&#x20;
+Simliarly, `cosmwasm-std` also has `Uint64` and `Uint256` and all of the following applies there as well.
+
+### When To Use Uint128 Instead Of u128
+
+`Uint128` is a thin wrapper around `u128` that is using strings for JSON encoding/decoding, such that the full `u128` range can be used for clients that convert JSON numbers to floats, like JavaScript and jq ([source](https://docs.cosmwasm.com/docs/1.0/smart-contracts/math/#uint128)).
```
Comment (eshelB) [inline development/secret-contracts/best-practices/data-handling.md:13]: I think "uses" sounds better than "is using"

File: development/secret-contracts/best-practices/data-handling.md [hunk 5]
```diff
@@ -2,14 +2,82 @@
...
+
+### Entrypoint Messages
+
+If you are familiar with Messages, you already know that most of the time we will use [serde](https://serde.rs/) to deserialize them from JSON (if not, you should read on [contract entrypoints](https://docs.cosmwasm.com/docs/1.0/actor-model/actor-in-blokchain/#entry-points) and the concept of [Messages](https://docs.cosmwasm.com/docs/1.0/smart-contracts/message/message)). Output will often be serizlized in the same way.
+
+In general, JSON implementations usually accept `[-(2^53)+1,(2^53)-1]` as an acceptable range for numbers. That's why we'll prefer to use `Uint128` in entrypoint messages, for example:
```
Comment (eshelB) [inline development/secret-contracts/best-practices/data-handling.md:19]: maybe strengthen connection between [range in json] to [reason to use Uint128]. e.g. "if we need more than (that/53 bits)"

Comment (toml01) [reply development/secret-contracts/best-practices/data-handling.md:19]: Can you give a full example? Not sure where to put it

File: development/secret-contracts/best-practices/data-handling.md (hunk 2 above)
Comment (toml01) [reply development/secret-contracts/best-practices/data-handling.md:67]: I feel like it's a bit too complex for this level of doc.
Do you have a simple example in mind?

**Comment by eshelB** [comment]
JSON implementations usually accept `[-(2^53)+1,(2^53)-1]` as an acceptable range for numbers. "That's why"(X) "If we need more than that"(V), we'll use `Uint128` in entrypoint messages, for example:

**Pull request merged**